modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/JsonFromTokenUri.swift

```diff
@@ -14,38 +14,47 @@ import Combine
 
 final class JsonFromTokenUri {
     private let tokensService: TokenProvidable
-    private lazy var getTokenUri = NonFungibleContract(blockchainProvider: blockchainProvider)
+    private let getTokenUri: NonFungibleContract
     private let blockchainProvider: BlockchainProvider
     private var inFlightPromises: [String: AnyPublisher<NonFungibleBalanceAndItsSource<JsonString>, SessionTaskError>] = [:]
     private let queue = DispatchQueue(label: "org.alphawallet.swift.jsonFromTokenUri")
     //Unlike `SessionManager.default`, this doesn't add default HTTP headers. It looks like POAP token URLs (e.g. https://api.poap.xyz/metadata/2503/278569) don't like them and return `406` in the JSON. It's strangely not responsible when curling, but only when running in the app
     private let networkService: NetworkService
 
-    public init(blockchainProvider: BlockchainProvider, tokensService: TokenProvidable, networkService: NetworkService) {
+    public init(blockchainProvider: BlockchainProvider,
+                tokensService: TokenProvidable,
+                networkService: NetworkService) {
+
         self.networkService = networkService
         self.blockchainProvider = blockchainProvider
         self.tokensService = tokensService
+        self.getTokenUri = NonFungibleContract(blockchainProvider: blockchainProvider)
+    }
+
+    deinit {
+        print("xxx.\(self).deinit for server: \(blockchainProvider.server)")
     }
 
     func fetchJsonFromTokenUri(forTokenId tokenId: String, tokenType: TokenType, address: AlphaWallet.Address, enjinToken: GetEnjinTokenQuery.Data.EnjinToken?) -> AnyPublisher<NonFungibleBalanceAndItsSource<JsonString>, SessionTaskError> {
         return Just(tokenId)
             .receive(on: queue)
             .setFailureType(to: SessionTaskError.self)
-            .flatMap { [queue, getTokenUri] tokenId -> AnyPublisher<NonFungibleBalanceAndItsSource<JsonString>, SessionTaskError> in
+            .flatMap { [weak self, queue, weak getTokenUri] tokenId -> AnyPublisher<NonFungibleBalanceAndItsSource<JsonString>, SessionTaskError> in
+                guard let strongSelf = self, let getTokenUri = getTokenUri else { return .fail(.cancelledError) }
                 let key = "\(tokenId).\(address.eip55String).\(tokenType.rawValue)"
 
-                if let promise = self.inFlightPromises[key] {
```

Comment:
Maybe not the best way to have accessed `self` previously, but `.flatMap()`'s `transform` arg is escaping, why would it retain `self` and cause a leak?

Reply:
but this sscaping closure captures self, and self captures publisher, that causes issue. we got strong ref to JsonFromTokenUri, in balacne fetcher, when we delete balance fetcher, e.g: switch rpc server, balance fetcher delates, (that is fine) but with it we want to delete all inner objects. so JsonFromTokenUri stays in memory

Reply:
flatMap from combine creates a new publisher. this escaping block captures self as well as self captures publisher

Reply:
😭 of course! I had read "escaping" as "non-escaping" (so I thought it should retain `self`). All good!